### ISSUE TITLE
Connection to storage account with storage account keys disabled doesn't work with workload identity (kerberos auth support)

### ISSUE THREAD

**Issue by Vegoo89** [kubernetes-sigs/azurefile-csi-driver#1737]
**What happened**: We mapped workload identity to UAMI with RBAC role (SMB Elevated Contributor) on desired storage account scope, however it raises permission denied on mount attempt when storage account keys are disabled on this storage account. When we enable storage account keys it starts to work again.

**What you expected to happen**: CSI Driver should use RBAC role instead of downloading storage keys 

**How to reproduce it**: Setup CSI connection to storage account with usage of workload identity -> disable storage keys on desired accounts -> try to run the application

**Anything else we need to know?**:

**Environment**:

- CSI Driver version: 1.30
- Kubernetes version (use `kubectl version`): 1.27.7
- OS (e.g. from /etc/os-release):  Ubuntu Linux
- Kernel (e.g. `uname -a`): managed by MS
- Install tools:
- Others:

**Comment by Vegoo89** [comment]
But can't it use RBAC role to authenticate? Is there any reason why it can't be done? My company is enrolling policy to disable shared keys so I am reviewing my options.

**Comment by Vegoo89** [comment]
Sorry if I am missing something but I am scratching my head now. We use workload identity on AKS for keyless auth to wide range of Azure resources. 

Can't we use it in similar way to authorize ourselves to file share if UAMI has required role assigned and is present under User Assigned identities on AKS VMSS?

**Comment by andyzhangx** [comment]
@Vegoo89 the azure file csi driver does not support keyless auth now unless you use NFS file share, that does not require key auth.

**Comment by Vegoo89** [comment]
I understand it is not supported now, however I want to understand what is the limitation and what would be necessary to work around it.

You told AKS nodes don't support AAD auth, but these are standard VMSS, managed by MS, right? If I assign UAMI to it, why can't I use it to authenticate to file share?

**Comment by andyzhangx** [comment]
in brief, there is no workaround for keyless auth since this driver only support key auth for smb mount unless you use NFS file share:
https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/3162e05a0036759c1313d31b83c830d3ab625be6/pkg/azurefile/nodeserver.go#L343

**Comment by Vegoo89** [comment]
Thanks a lot for detailed explanation. May I keep it opened until smb mount supports AAD or it won't happen in nearest future?

**Comment by djsly** [comment]
@andyzhangx  am I understanding that with NFS, we can disable Key Access ?  We are also being asked to disable `Allow storage account key access` on ALL Storage Account.  

1- I don't see a option for disabling that even if using NFS correct ?

**Comment by andyzhangx** [comment]
> @andyzhangx am I understanding that with NFS, we can disable Key Access ? We are also being asked to disable `Allow storage account key access` on ALL Storage Account.
> 
> 1- I don't see a option for disabling that even if using NFS correct ?

@djsly yes, you could disable account key access if you are only using nfs file share. We will add such options for account created by this driver.

**Comment by rohank07** [comment]
@andyzhangx Any updates on this?

**Comment by andyzhangx** [comment]
> [@andyzhangx](https://github.com/andyzhangx) Any updates on this?

@rohank07 we are working on managed identity auth mount in phase 1: https://github.com/kubernetes-sigs/azurefile-csi-driver/pull/2652, and will work on workload identity auth mount in phase 2.

**Comment by andyzhangx** [comment]
mount with managed identity (without using account key) is already supported on AKS 1.34, check guide here: https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/master/docs/managed-identity-mount.md, and we are still working on mounting with workload identity

**Comment by Johno-ACSLive** [comment]
@andyzhangx I'm getting the below error on AKS 1.34 (note I've replaced some text with RPL) "message: >-
  MountVolume.MountDevice failed for volume "pv-prd-RPL" : rpc error: code =
  Internal desc = volume(RPL) mount
  //RPL.file.core.windows.net/RPL on
  /var/lib/kubelet/plugins/kubernetes.io/csi/file.csi.azure.com/RPL/globalmount
  failed with setCredentialCache failed for RPL.file.core.windows.net
  with error: exit status 1, output: [-] Error calling AzAuthenticatorLib: -1

  Error getting Kerberos service ticket, check /var/log/syslog for more
  information."

/var/log/syslog doesn't show any other meaninful information and logging level was increased (to 5 from memory).

I've got a ticket open with Microsoft support on this. AKS node can connect to the storage account using account key, just not using the managed identity. Permissions all setup correctly (managed identity of node agent configured for "Storage File Data SMB MI Admin") and Storage Account has Identity-based access enabled using Microsoft Entra Kerberos.

**Comment by andyzhangx** [comment]
> [@andyzhangx](https://github.com/andyzhangx) I'm getting the below error on AKS 1.34 (note I've replaced some text with RPL) "message: >- MountVolume.MountDevice failed for volume "pv-prd-RPL" : rpc error: code = Internal desc = volume(RPL) mount //RPL.file.core.windows.net/RPL on /var/lib/kubelet/plugins/kubernetes.io/csi/file.csi.azure.com/RPL/globalmount failed with setCredentialCache failed for RPL.file.core.windows.net with error: exit status 1, output: [-] Error calling AzAuthenticatorLib: -1
> 
> Error getting Kerberos service ticket, check /var/log/syslog for more information."
> 
> /var/log/syslog doesn't show any other meaninful information and logging level was increased (to 5 from memory).
> 
> I've got a ticket open with Microsoft support on this. AKS node can connect to the storage account using account key, just not using the managed identity. Permissions all setup correctly (managed identity of node agent configured for "Storage File Data SMB MI Admin") and Storage Account has Identity-based access enabled using Microsoft Entra Kerberos.

@Johno-ACSLive If you are using your own storage account, please ensure that the `SMBOauth` property is enabled for that account.
```console
Set-AzStorageAccount -ResourceGroupName MyRG -Name expshaccount -EnableSmbOAuth $true 
```

**Comment by Johno-ACSLive** [comment]
@andyzhangx thank you, I see that mention of SmbOAuth was added recently to the managed-identity-mount.md file!

**Comment by RomanBednar** [comment]
> we are still working on mounting with workload identity

Hi @andyzhangx, I don’t mean to be pushy, but I’m wondering if you have any estimate for when this will ship. We’re trying to determine whether it has any chance to land in OpenShift 4.22.

**Comment by andyzhangx** [comment]
> > we are still working on mounting with workload identity
> 
> Hi [@andyzhangx](https://github.com/andyzhangx), I don’t mean to be pushy, but I’m wondering if you have any estimate for when this will ship. We’re trying to determine whether it has any chance to land in OpenShift 4.22.

@RomanBednar Sorry I cannot provide you an ETA since the dependency is on smb kernel driver side, Azure file team is investigating how to support mounting with workload identity, and currently mounting with managed identity is already supported in CSI driver v1.34.0

**Comment by vivianzheng404** [comment]
Storage account key is not the most secure way. Would workload identity be supported in multiple tenants? I think the solution proposed in https://github.com/kubernetes-sigs/azurefile-csi-driver/issues/2131 is worth exploring. 

There are at least 3 issues raised regarding multiple tenants workload identity support. This feature is highly desired by the community and has real world values.

https://github.com/kubernetes-sigs/azurefile-csi-driver/issues/1737
https://github.com/kubernetes-sigs/azurefile-csi-driver/issues/2131
https://github.com/kubernetes-sigs/azurefile-csi-driver/issues/2393

**Comment by andyzhangx** [comment]
workload identity is already supported on AKS: https://github.com/kubernetes-sigs/azurefile-csi-driver/blob/master/docs/workload-identity-static-pv-mount.md
